nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java

```diff
@@ -1085,6 +1087,28 @@ private static Object toEnum(Object value, EnumDataType dataType, String fieldNa
         throw new IllegalTypeConversionException("Cannot convert value [" + value + "] of type " + value.getClass() + " to Date for field " + fieldName);
     }
 
+    /**
+     * Converts a java.sql.Date object with 00:00:00 in local time zone (typically coming from a java.sql.ResultSet)
+     * to UTC normalized form (storing epoch corresponding to UTC 00:00:00 on the given day).
+     *
+     * @param date java.sql.Date with local time zone 00:00:00
+     * @return java.sql.Date with UTC 00:00:00
+     */
+    public static Date convertDateToUTC(Date date) {
+        return new Date(date.toLocalDate().atStartOfDay(ZoneId.of("UTC")).toInstant().toEpochMilli());
+    }
+
+    /**
+     * Converts a java.sql.Date object with 00:00:00 in UTC
+     * to local time zone normalized form (storing epoch corresponding to 00:00:00 in local time zone on the given day).
+     *
+     * @param date java.sql.Date with UTC 00:00:00
+     * @return java.sql.Date with local time zone 00:00:00
+     */
+    public static Date convertDateToLocalTZ(Date date) {
+        return Date.valueOf(Instant.ofEpochMilli(date.getTime()).atZone(ZoneId.of("UTC")).toLocalDate());
```

Comment:
The method naming and implementation are a little confusing on initial read.  The `atZone()` method returns a `ZonedDateTime` and `toLocalDate()` return the `java.time.LocalDate` portion, but does that actually convert the date to the system default time zone?  It may be helpful to break this into multiple lines as it is a bit hard to follow.

Reply:
Indeed... what do you think about this?
```
    public static Date convertDateToUTC(Date dateLocalTZ) {
        ZonedDateTime zdtLocalTZ = ZonedDateTime.ofInstant(Instant.ofEpochMilli(dateLocalTZ.getTime()), ZoneId.systemDefault());
        ZonedDateTime zdtUTC = zdtLocalTZ.withZoneSameLocal(ZoneId.of("UTC"));
        Date dateUTC = new Date(zdtUTC.toInstant().toEpochMilli());
        return dateUTC;
    }
```
No `LocaDate` and I believe `withZoneSameLocal()` in the middle of the method really grasps what is going on here.

Reply:
Thanks for the updated suggestion, breaking things out read much more easily, and the `withZoneSameLocal()` method makes more sense now.  Based on that snippet, I would suggest just returning the final `Date` object instead of declaring and returning.